proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -371,18 +374,393 @@ lemma decode_sc_ctrl_inv_corres:
              split: cap.splits)
   done
 
-(* FIXME RT: preconditions can be reduced, this is what is available at the call site: *)
-lemma invoke_sched_context_corres:
+lemma schedContextBindNtfn_corres:
+  "corres dc
+     (valid_objs and sc_ntfn_sc_at ((=) None) scp
+      and (obj_at (\<lambda>ko. \<exists>ntfn. ko = Notification ntfn \<and> ntfn_sc ntfn = None) ntfnp))
+     (ntfn_at' ntfnp and sc_at' scp)
+     (sched_context_bind_ntfn scp ntfnp) (schedContextBindNtfn scp ntfnp)"
+  unfolding sched_context_bind_ntfn_def schedContextBindNtfn_def
+  apply (clarsimp simp: update_sk_obj_ref_def bind_assoc)
+  apply (rule corres_guard_imp)
+    apply (rule corres_split[OF getNotification_corres])
+      apply (rule corres_guard_imp)
+        apply (rule corres_split[OF setNotification_corres])
+           apply (clarsimp simp: ntfn_relation_def split: Structures_A.ntfn.splits)
+          apply (fold updateSchedContext_def)
+          apply (rule updateSchedContext_corres)
+            apply (clarsimp simp: opt_map_red obj_at_simps is_sc_obj)
+            apply (drule (1) pspace_relation_absD[OF _ state_relation_pspace_relation])
+            apply (clarsimp simp: sc_relation_def)
+           apply (clarsimp simp: opt_map_red obj_at_simps is_sc_obj)
+           apply (drule (1) sc_replies_relation_prevs_list'[OF state_relation_sc_replies_relation])
+           apply (clarsimp simp: opt_map_red)
+          apply (clarsimp simp: objBits_simps')
+         apply wpsimp
+        apply wpsimp
+       apply simp+
+     apply wpsimp+
+   apply (fastforce simp: obj_at_simps sc_ntfn_sc_at_def is_ntfn is_sc_obj valid_obj_def)
+  apply clarsimp
+  done
+
+crunches tcb_sched_action, complete_yield_to, reschedule_required, sched_context_resume
+  for in_user_frame[wp]: "in_user_frame buf"
+  (simp: crunch_simps wp: crunch_wps ignore: set_tcb_obj_ref)
+
+lemma
```

Comment:
These can't be crunched?

Reply:
I guess not, if you do some crunching immediately after. I wonder if it's worth investigating

Reply:
I did try initially and didn’t work.